api/controlplane/kubeadm/v1beta2/kubeadm_control_plane_types.go

```diff
@@ -540,31 +524,57 @@ type KubeadmControlPlaneMachineTemplateDeletionSpec struct {
 	NodeDeletionTimeoutSeconds *int32 `json:"nodeDeletionTimeoutSeconds,omitempty"`
 }
 
-// RolloutBefore describes when a rollout should be performed on the KCP machines.
-type RolloutBefore struct {
+// KubeadmControlPlaneRolloutSpec defines the rollout behavior.
+// +kubebuilder:validation:MinProperties=1
+type KubeadmControlPlaneRolloutSpec struct {
+	// before is a field to indicate a rollout should be performed
+	// if the specified criteria is met.
+	// +optional
+	Before KubeadmControlPlaneRolloutBeforeSpec `json:"before,omitempty,omitzero"`
+
+	// after is a field to indicate a rollout should be performed
+	// after the specified time even if no changes have been made to the
+	// KubeadmControlPlane.
+	// Example: In the YAML the time can be specified in the RFC3339 format.
+	// To specify the rolloutAfter target as March 9, 2023, at 9 am UTC
+	// use "2023-03-09T09:00:00Z".
+	// +optional
+	After *metav1.Time `json:"after,omitempty"`
+
+	// strategy specifies how to roll out control plane Machines.
+	// +optional
+	Strategy KubeadmControlPlaneRolloutStrategy `json:"strategy,omitempty,omitzero"`
+}
+
+// KubeadmControlPlaneRolloutBeforeSpec describes when a rollout should be performed on the KCP machines.
+// +kubebuilder:validation:MinProperties=1
+type KubeadmControlPlaneRolloutBeforeSpec struct {
 	// certificatesExpiryDays indicates a rollout needs to be performed if the
 	// certificates of the machine will expire within the specified days.
 	// +optional
-	CertificatesExpiryDays *int32 `json:"certificatesExpiryDays,omitempty"`
+	// +kubebuilder:validation:Minimum=7
+	CertificatesExpiryDays int32 `json:"certificatesExpiryDays,omitempty"`
 }
 
-// RolloutStrategy describes how to replace existing machines
+// KubeadmControlPlaneRolloutStrategy describes how to replace existing machines
 // with new ones.
-type RolloutStrategy struct {
+// +kubebuilder:validation:MinProperties=1
+type KubeadmControlPlaneRolloutStrategy struct {
```

Comment:
CEL can very easily validate the relationship between `type` and `rollingUpdate` here
```
+kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'RollingUpdate' ? has(self.rollingUpdate) : !has(self.rollingUpdate)",message="rollingUpdate is required when type is RollingUpdate, and forbidden otherwise"
```

Reply:
> rollingUpdate is required when type is RollingUpdat

This is not correct. rollingUpdate is not required when type = RollingUpdate